electrolytes/solvation_shell_utils.py

```diff
@@ -0,0 +1,152 @@
+import re
+import numpy as np
+import random
+from typing import List, Set, Callable
+
+from schrodinger.structutils import analyze
+from schrodinger.application.matsci import clusterstruct
+from schrodinger.comparison import are_conformers
+from schrodinger.application.jaguar.utils import group_with_comparison
+from schrodinger.comparison.atom_mapper import ConnectivityAtomMapper
+from schrodinger.structutils import rmsd
+from schrodinger.structure import Structure
+
+
+def filter_shells_with_solute_atoms(
+    shells: List[Set[int]], st: Structure, solute_res_names: List[str]
+) -> List[Structure]:
+    """
+    Filter out shells that contain solute atoms.
+    Args:
+        shells: List of sets of atom indices that correspond to shells
+        st: Entire structure from the PDB file
+        solute_res_names: List of residue names that correspond to solute atoms in the simulation
+    Returns:
+        List of Schrodinger structures that do not contain solute atoms
+    """
+    actual_shells = [st.extract(shell, copy_props=True) for shell in shells]
+    filtered_shells = [
+        shell
+        for shell in actual_shells
+        if not any(
+            at.getResidue().pdbres.strip() in solute_res_names for at in shell.atom
+        )
+    ]
+    return filtered_shells
+
+
+def expand_shell(
+    st: Structure,
+    shell_ats: Set[int],
+    central_solute: int,
+    radius: float,
+    solute_res_names: List[str],
+    max_shell_size: int = 200,
+) -> Structure:
+    """
+    Expands a solvation shell. If there are any (non-central) solutes present in the shell,
+    recursively include shells around those solutes.
+    First, gets the molecule numbers of solute molecules that are within the radius
+    and not already expanded around. Then, continuously expand around them as long as we don't hit an atom limit.
+    Args:
+        st: Entire structure from the PDB file
+        shell_ats: Set of atoms in a shell (1-indexed)
+        central_solute: Index of the central solute atom in the shell
+        radius: Solvation radius (Angstroms) to consider
+        solute_res_names: List of residue names that correspond to solute atoms in the simulation
+        max_shell_size: Maximum size (in atoms) of the expanded shell
+    Returns:
+        Structure object containing the expanded solvation shell
+    """
+    # TODO: I'm not sure why we have to do this, but sometimes the central solute is not in the shell
+    if central_solute not in shell_ats:
+        shell_ats.add(central_solute)
+
+    solutes_included = set([central_solute])
+    while len(shell_ats) < max_shell_size:
```

Comment:
Note that the maximum size is a little uncontrolled with this logic. We exit the while loop if we don't have any more solutes to expand around (obviously and correct) or when the number of atoms is greater than or equal to max_shell_size (a bit fuzzy).

Reply:
I think it's fine either way. If we put the "fuzzy" cap lower, with the understanding that we'll have some structures that are actually somewhat larger, it'll do basically the same job as having a higher, hard cap.

Reply:
Yeah I agree that it's a bit fuzzy. I've made a small modification to the loop to store the previous `shell_ats` object right before the size was exceeded and then return that. So now the cap is hard.